File: src/types/doctor.ts
// src/types/doctor.ts

export interface Doctor {
  id: string;
  email: string;
  password?: string;
  firstName: string;
  lastName: string;
  fullName: string;
  phone: string;
  dateOfBirth: string;
  gender: 'male' | 'female' | 'other';

  // Professional Info
  specialization: string;
  subSpecializations: string[];
  qualifications: Qualification[];
  experience: number; // years
  licenseNumber: string;
  medicalCouncil: string;

  // Hospital Info
  hospitalName: string;
  hospitalAddress: Address;
  department: string;
  designation: string;

  // Consultation
  consultationFee: number;
  followUpFee: number;
  videoConsultationFee: number;
  consultationDuration: number; // minutes
  languages: string[];
  consultationModes: ('in-person' | 'video' | 'phone')[];

  // Availability
  availabilitySchedule: WeeklySchedule;
  isOnline: boolean;
  isAvailable: boolean;
  maxPatientsPerDay: number;

  // Stats
  rating: number;
  reviewCount: number;
  totalPatients: number;
  totalConsultations: number;
  successRate: number;
  experience_years: number;

  // Profile
  about: string;
  profileImage: string;
  coverImage: string;
  achievements: string[];
  awards: Award[];
  publications: Publication[];
  memberships: string[];

  // Contact
  emailVerified: boolean;
  phoneVerified: boolean;
  isActive: boolean;
  joinedDate: string;

  // Settings
  notificationsEnabled: boolean;
  smsAlertsEnabled: boolean;
  emailAlertsEnabled: boolean;

  address: string;
  education: Education[];
  status: 'online' | 'offline' | 'busy';
}

export interface Qualification {
  degree: string;
  institution: string;
  year: number;
  country: string;
}

export interface Education {
  degree: string;
  institution: string;
  year: number;
}

export interface Award {
  title: string;
  organization: string;
  year: number;
  description?: string;
}

export interface Publication {
  title: string;
  journal: string;
  year: number;
  doi?: string;
  citations?: number;
}

export interface Address {
  street: string;
  city: string;
  state: string;
  zipCode: string;
  country: string;
  coordinates?: {
    latitude: number;
    longitude: number;
  };
}

export interface WeeklySchedule {
  monday: DaySchedule;
  tuesday: DaySchedule;
  wednesday: DaySchedule;
  thursday: DaySchedule;
  friday: DaySchedule;
  saturday: DaySchedule;
  sunday: DaySchedule;
}

export interface DaySchedule {
  isAvailable: boolean;
  slots: TimeSlot[];
}

export interface TimeSlot {
  startTime: string;
  endTime: string;
  maxPatients: number;
  currentPatients: number;
  isAvailable: boolean;
}

export interface Patient {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  dateOfBirth: string;
  bloodGroup: BloodGroup;
  gender: string;
  lastVisit: string;
  totalVisits: number;
  upcomingAppointment?: Appointment;
  medicalHistory: PatientMedicalHistory;
  currentMedications: Medication[];
  allergies: string[];
  notes?: string;
}

export type BloodGroup = 'A+' | 'A-' | 'B+' | 'B-' | 'AB+' | 'AB-' | 'O+' | 'O-';

export interface PatientMedicalHistory {
  conditions: MedicalCondition[];
  surgeries: Surgery[];
  familyHistory: string[];
}

export interface MedicalCondition {
  name: string;
  diagnosedDate: string;
  status: 'active' | 'resolved' | 'managed';
  notes?: string;
}

export interface Surgery {
  name: string;
  date: string;
  hospital: string;
  notes?: string;
}

export interface Medication {
  id: string;
  name: string;
  dosage: string;
  frequency: string;
  startDate: string;
  endDate?: string;
  prescribedBy: string;
}

export interface Appointment {
  id: string;
  patientId: string;
  patientName: string;
  patientPhone: string;
  patientEmail: string;
  date: string;
  time: string;
  duration: number;
  type: 'in-person' | 'video' | 'phone';
  status: 'scheduled' | 'confirmed' | 'in-progress' | 'completed' | 'cancelled' | 'no-show';
  reason: string;
  symptoms?: string[];
  isEmergency: boolean;
  isFirstVisit: boolean;
  notes?: string;
  prescriptionId?: string;
  followUpDate?: string;
  createdAt: string;
  updatedAt: string;
}

export interface Prescription {
  id: string;
  patientId: string;
  patientName: string;
  doctorId: string;
  doctorName: string;
  appointmentId: string;
  date: string;
  diagnosis: string;
  symptoms: string[];
  medications: PrescribedMedication[];
  tests: PrescribedTest[];
  advice: string;
  followUpDate?: string;
  validUntil: string;
  isDigital: boolean;
  digitalSignature?: string;
  status: 'active' | 'completed' | 'expired';
}

export interface PrescribedMedication {
  name: string;
  dosage: string;
  frequency: string;
  duration: string;
  timing: 'before-food' | 'after-food' | 'with-food' | 'empty-stomach';
  quantity: number;
  refills: number;
  instructions?: string;
}

export interface PrescribedTest {
  name: string;
  type: string;
  instructions?: string;
  isUrgent: boolean;
}

export interface BloodRequest {
  id: string;
  patientId: string;
  patientName: string;
  bloodGroup: BloodGroup;
  units: number;
  hospital: string;
  urgency: 'normal' | 'urgent' | 'emergency';
  status: 'pending' | 'approved' | 'rejected' | 'fulfilled';
  requestDate: string;
  requiredDate: string;
  reason: string;
  approvedBy?: string;
  approvalDate?: string;
}

export interface VideoConsultation {
  id: string;
  appointmentId: string;
  doctorId: string;
  patientId: string;
  patientName: string;
  date: string;
  startTime: string;
  endTime?: string;
  status: 'scheduled' | 'waiting' | 'in-progress' | 'completed' | 'missed';
  roomId: string;
  recordingUrl?: string;
  notes?: string;
}

export interface DoctorDashboardData {
  todayAppointments: Appointment[];
  upcomingAppointments: Appointment[];
  emergencyRequests: BloodRequest[];
  pendingPrescriptions: Prescription[];
  videoConsultations: VideoConsultation[];
  recentPatients: Patient[];
  earnings: EarningsData;
  stats: DoctorStats;
  activities: Activity[];
  notifications: DoctorNotification[];
}

export interface EarningsData {
  today: number;
  thisWeek: number;
  thisMonth: number;
  total: number;
  breakdown: {
    consultations: number;
    videoConsultations: number;
    followUps: number;
  };
  chartData: {
    labels: string[];
    values: number[];
  };
}

export interface DoctorStats {
  totalPatients: number;
  todayPatients: number;
  completedAppointments: number;
  cancelledAppointments: number;
  averageRating: number;
  totalReviews: number;
  prescriptionCount: number;
  videoConsultCount: number;
}

export interface Activity {
  id: string;
  type: 'appointment' | 'prescription' | 'consultation' | 'review' | 'emergency';
  description: string;
  patientName?: string;
  time: string;
  status: 'completed' | 'pending' | 'cancelled';
}

export interface DoctorNotification {
  id: string;
  type: 'appointment' | 'emergency' | 'prescription' | 'review' | 'system';
  title: string;
  message: string;
  isRead: boolean;
  createdAt: string;
  actionUrl?: string;
  priority: 'low' | 'medium' | 'high' | 'urgent';
}

File: backend/app/schemas/doctor.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import date, datetime


# ============================================
# DOCTOR PROFILE SCHEMAS
# ============================================
class DoctorProfileResponse(BaseModel):
    id: str
    user_id: str
    specialization: str
    sub_specializations: Optional[List[str]] = None
    license_number: Optional[str] = None
    medical_council: Optional[str] = None
    experience_years: int = 0
    qualifications: Optional[List] = None
    consultation_fee: float = 0
    follow_up_fee: float = 0
    video_consultation_fee: float = 0
    hospital_affiliation: Optional[str] = None
    department: Optional[str] = None
    designation: Optional[str] = None
    languages: Optional[List[str]] = None
    is_verified: bool = False
    rating: float = 0.0
    review_count: int = 0
    total_patients: int = 0
    total_consultations: int = 0
    success_rate: float = 0.0
    about: Optional[str] = None
    profile_image: Optional[str] = None
    cover_image: Optional[str] = None
    achievements: Optional[List] = None
    awards: Optional[List] = None
    publications: Optional[List] = None
    memberships: Optional[List] = None
    status: str = "offline"
    is_online: bool = False
    is_active: bool = True
    max_patients_per_day: int = 30
    consultation_duration: int = 15
    consultation_modes: Optional[List] = None
    # User fields joined
    full_name: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    gender: Optional[str] = None
    date_of_birth: Optional[date] = None
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class DoctorProfileUpdateRequest(BaseModel):
    specialization: Optional[str] = Field(None, max_length=150)
    sub_specializations: Optional[List[str]] = None
    medical_council: Optional[str] = Field(None, max_length=150)
    experience_years: Optional[int] = Field(None, ge=0)
    qualifications: Optional[List] = None
    consultation_fee: Optional[float] = Field(None, ge=0)
    follow_up_fee: Optional[float] = Field(None, ge=0)
    video_consultation_fee: Optional[float] = Field(None, ge=0)
    hospital_affiliation: Optional[str] = Field(None, max_length=255)
    department: Optional[str] = Field(None, max_length=100)
    designation: Optional[str] = Field(None, max_length=100)
    languages: Optional[List[str]] = None
    about: Optional[str] = None
    achievements: Optional[List] = None
    awards: Optional[List] = None
    publications: Optional[List] = None
    memberships: Optional[List] = None
    max_patients_per_day: Optional[int] = Field(None, ge=1, le=100)
    consultation_duration: Optional[int] = Field(None, ge=5, le=120)
    consultation_modes: Optional[List[str]] = None


# ============================================
# DOCTOR LIST ITEM (for search/browse)
# ============================================
class DoctorListItemResponse(BaseModel):
    id: str
    user_id: str
    specialization: str
    experience_years: int = 0
    consultation_fee: float = 0
    hospital_affiliation: Optional[str] = None
    rating: float = 0.0
    review_count: int = 0
    total_patients: int = 0
    about: Optional[str] = None
    profile_image: Optional[str] = None
    status: str = "offline"
    is_online: bool = False
    is_verified: bool = False
    # User fields
    full_name: Optional[str] = None
    gender: Optional[str] = None
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ============================================
# DOCTOR AVAILABILITY
# ============================================
class DoctorAvailabilityResponse(BaseModel):
    id: str
    doctor_id: str
    day_of_week: int
    start_time: str
    end_time: str
    max_patients: int = 10
    current_patients: int = 0
    is_available: bool = True

    model_config = {"from_attributes": True}


class DoctorAvailabilityCreateRequest(BaseModel):
    day_of_week: int = Field(..., ge=0, le=6)
    start_time: str = Field(..., max_length=10)
    end_time: str = Field(..., max_length=10)
    max_patients: int = Field(default=10, ge=1, le=100)
    is_available: bool = True


class WeeklyScheduleUpdateRequest(BaseModel):
    schedule: List[DoctorAvailabilityCreateRequest]


# ============================================
# DOCTOR DASHBOARD
# ============================================
class DoctorDashboardResponse(BaseModel):
    today_appointments: int = 0
    upcoming_appointments: int = 0
    total_patients: int = 0
    total_consultations: int = 0
    completed_appointments: int = 0
    cancelled_appointments: int = 0
    average_rating: float = 0.0
    total_reviews: int = 0
    today_earnings: float = 0.0
    this_week_earnings: float = 0.0
    this_month_earnings: float = 0.0
    total_earnings: float = 0.0
    recent_activities: List[dict] = []
    notifications: List[dict] = []


# ============================================
# DOCTOR PATIENT (relationship)
# ============================================
class DoctorPatientResponse(BaseModel):
    id: str
    doctor_id: str
    patient_id: str
    total_visits: int = 0
    last_visit: Optional[date] = None
    is_active: bool = True
    # Joined patient info
    full_name: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    gender: Optional[str] = None
    profile_image: Optional[str] = None
    blood_group: Optional[str] = None

    model_config = {"from_attributes": True}


# ============================================
# DOCTOR NOTIFICATION
# ============================================
class DoctorNotificationResponse(BaseModel):
    id: str
    doctor_id: str
    type: str
    title: str
    message: str
    is_read: bool = False
    action_url: Optional[str] = None
    priority: str = "medium"
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ============================================
# DOCTOR EARNINGS
# ============================================
class DoctorEarningsResponse(BaseModel):
    id: str
    doctor_id: str
    earning_date: date
    consultations: int = 0
    video_consultations: int = 0
    follow_ups: int = 0
    total_amount: float = 0

    model_config = {"from_attributes": True}


# ============================================
# DOCTOR ACTIVITY
# ============================================
class DoctorActivityResponse(BaseModel):
    id: str
    doctor_id: str
    type: str
    description: str
    patient_name: Optional[str] = None
    status: str = "completed"
    created_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ============================================
# DOCTOR PROFILE CREATE REQUEST
# ============================================
class DoctorProfileCreateRequest(BaseModel):
    """Schema for creating a new doctor profile during registration."""
    specialization: str = Field(..., min_length=1, max_length=150)
    sub_specializations: Optional[List[str]] = None
    license_number: str = Field(..., min_length=1, max_length=100)
    medical_council: Optional[str] = None
    experience_years: int = Field(default=0, ge=0)
    qualifications: List[dict] = Field(default_factory=list)
    consultation_fee: float = Field(default=0, ge=0)
    follow_up_fee: float = Field(default=0, ge=0)
    video_consultation_fee: float = Field(default=0, ge=0)
    hospital_affiliation: Optional[str] = None
    department: Optional[str] = None
    designation: Optional[str] = None
    languages: Optional[List[str]] = None
    consultation_duration: int = Field(default=15, ge=5, le=120)
    max_patients_per_day: int = Field(default=30, ge=1, le=100)
    consultation_modes: Optional[List[str]] = None
    about: Optional[str] = None
    achievements: Optional[List[str]] = None
    awards: Optional[List[dict]] = None
    publications: Optional[List[dict]] = None
    memberships: Optional[List[str]] = None
